packages/grid/x-data-grid/src/hooks/features/columnHeaders/useGridColumnHeaders.tsx

```diff
@@ -268,18 +300,210 @@ export const useGridColumnHeaders = (props: UseGridColumnHeadersProps) => {
+
+      const depthInfo = renderedColumns.reduce((aggregated, column, i) => {
+        const lastItem: HeaderInfo | undefined = aggregated[aggregated.length - 1];
+
+        if (column.groupPath && column.groupPath.length > depth) {
+          if (lastItem && lastItem.groupId === column.groupPath[depth]) {
+            // Merge with the previous columns
+            return [
+              ...aggregated.slice(0, aggregated.length - 1),
+              {
+                ...lastItem,
+                width: lastItem.width + (column.width ?? 0),
+                fields: [...lastItem.fields, column.field],
+              },
+            ];
+          }
+          // Create a new grouping
+          return [
+            ...aggregated,
+            {
+              groupId: column.groupPath[depth],
+              width: column.width ?? 0,
+              fields: [column.field],
+              colIndex: firstColumnToRender + i,
+            },
+          ];
+        }
+
+        // It is the first level for which their is no group
+        if (MERGE_EMPTY_CELLS && lastItem && lastItem.groupId === null) {
+          // We merge with previous column
+          return [
+            ...aggregated.slice(0, aggregated.length - 1),
+            {
+              ...lastItem,
+              width: lastItem.width + (column.width ?? 0),
+              fields: [...lastItem.fields, column.field],
+            },
+          ];
+        }
+        // We create new empty cell
+        return [
+          ...aggregated,
+          {
+            groupId: null,
+            width: column.width ?? 0,
+            fields: [column.field],
+            colIndex: firstColumnToRender + i,
+          },
+        ];
+      }, initialHeader);
+
+      columnIndex = lastColumnToRender;
+      const lastColumnToRenderGroup = depthInfo[depthInfo.length - 1].groupId;
+
+      while (
+        lastColumnToRenderGroup !== null &&
+        columnIndex < maxLastColumn &&
+        visibleColumns[columnIndex]?.groupPath &&
+        visibleColumns[columnIndex]?.groupPath?.[depth] === lastColumnToRenderGroup
+      ) {
+        const column = visibleColumns[columnIndex];
+
+        depthInfo[depthInfo.length - 1].width += column.width ?? 0;
+        depthInfo[depthInfo.length - 1].fields.push(column.field);
+        columnIndex += 1;
+      }
+
+      headerToRender.push({ leftOverflow, elements: [...depthInfo] });
+    }
+
+    headerToRender.forEach((depthInfo, depthIndex) => {
+      columns.push(
+        <GridColumnHeaderRow
+          style={{
+            height: `${headerGroupingRowHeight}px`,
+            lineHeight: `${headerGroupingRowHeight}px`,
+            transform: `translateX(-${depthInfo.leftOverflow}px)`,
+          }}
+          key={depthIndex}
+          role="row"
+          aria-rowindex={depthIndex + 1}
+        >
+          {depthInfo.elements.map(({ groupId, width, fields, colIndex }, groupIndex) => {
+            return (
+              <GridColumnGroupHeader
```

Comment:
Could `GridColumnHeaderItem` be reused to have the same styling? We could add props to make it not clickable.

Reply:
It's possible but it will be a bit hacky, because `GridColumnHeaderItem`infers a lot of properties from `column` so I will have to provide a `column` props faking the column state from the parameters of the grouping model.

I'm also concerned by the number of conditions to introduce in `GridColumnHeaderItem` to support the two types of headers.

---
For example `data-field` becomes `data-fields` to be able to have a smooth resizing

The resizing function of a group will probably be different from the resizing of a column because for a column on each `mouseMove` we can get the difference before/after and apply the `2px` modification to all cells.
But when a column group containing 4 columns is resized by `1px` I'm not sure we can modify all rows by 0.25px

When introducing a collapsed/extended cell it will be new action buttons to set

---

Could it be possible to keep `GridColumnHeaderItem` and `GridColumnGroupHeaderItem` with their own interfaces and internal logic, but reuse the same styled-components in both such that we kept a unified style, but have custom props/logic adapted for each case

Reply:
I think the current approach is good. I'm most concerned about reusing the same class names.